### PR TITLE
[SPARK-32320][PYSPARK] Remove mutable default arguments

### PR DESCRIPTION
**Pull request by Fokko**
This is bad practice, and might lead to unexpected behaviour:
https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

```
fokkodriesprong@Fan spark % grep -R "={}" python | grep def

python/pyspark/resource/profile.py:    def __init__(self, _java_resource_profile=None, _exec_req={}, _task_req={}):
python/pyspark/sql/functions.py:def from_json(col, schema, options={}):
python/pyspark/sql/functions.py:def to_json(col, options={}):
python/pyspark/sql/functions.py:def schema_of_json(json, options={}):
python/pyspark/sql/functions.py:def schema_of_csv(csv, options={}):
python/pyspark/sql/functions.py:def to_csv(col, options={}):
python/pyspark/sql/functions.py:def from_csv(col, schema, options={}):
python/pyspark/sql/avro/functions.py:def from_avro(data, jsonFormatSchema, options={}):
```

```
fokkodriesprong@Fan spark % grep -R "=\[\]" python | grep def
python/pyspark/ml/tuning.py:    def __init__(self, bestModel, avgMetrics=[], subModels=None):
python/pyspark/ml/tuning.py:    def __init__(self, bestModel, validationMetrics=[], subModels=None):
```

### What changes were proposed in this pull request?

Removing the mutable default arguments.

### Why are the changes needed?

Removing the mutable default arguments, and changing the signature to `Optional[...]`.

### Does this PR introduce _any_ user-facing change?

No 👍 

### How was this patch tested?

Using the Flake8 bugbear code analysis plugin.

### CODE REVIEW

File: python/pyspark/sql/functions.py
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:2396]: I am happy with trying Python type hints. Actually, I am thinking about somehow leveraging (or porting?) https://github.com/zero323/pyspark-stubs, cc @zero323. Let's avoid to introduce the type hinting in the main source for now because we could decouple and run it separately.

Comment (Fokko) [reply python/pyspark/sql/functions.py:2396]: While touching the lines anyway, I thought is would be a nice addition. Having it in the source also makes live easier for the developers that work on pyspark itself. Curious what Maciej thinks.

Comment (zero323) [reply python/pyspark/sql/functions.py:2396]: > Furthermore, it is easier to keep the code and annotations in sync. Curious what Maciej thinks.

That's the topic for a long discussion :) Overall, I made some points [here](http://apache-spark-developers-list.1001551.n3.nabble.com/Re-PySpark-Revisiting-PySpark-type-annotations-td26232.html) and [here](http://apache-spark-developers-list.1001551.n3.nabble.com/PYTHON-PySpark-typing-hints-td21560.html) ‒ I think these are still valid.

Now... I am still opened about merging `pyspark-stubs` about with the main project. However, have some annotations in the code-base, if there is no will to maintain these (and maintaining useful annotations is honestly a pain) and bring them to high coverage any time soon, as it may make things either wasted effort (if not used at all, if package is not PEP 561 compliant) or maybe even harmful. That's a topic for a longer discussion though...

In general I'd bring the discussion to dev and or JIRA first.

File: python/pyspark/ml/tuning.py
Comment (zero323) [inline python/pyspark/ml/tuning.py:504]: Out of curiosity ‒why do we need default here?

File: python/pyspark/sql/functions.py
Comment (ueshin) [inline python/pyspark/sql/functions.py:2445]: We should use `None` here as well?

File: python/pyspark/ml/tuning.py
Comment (Fokko) [reply python/pyspark/ml/tuning.py:504]: Good question, I don't think these should be optional

File: python/pyspark/sql/functions.py
Comment (Fokko) [reply python/pyspark/sql/functions.py:2396]: Thanks for pointing out the maillist. Python <3.6 has been dropped in the meantime https://github.com/apache/spark/pull/28957.

I'm happy to help with migrating the annotations to the codebase. I still believe that it makes much sense to keep them sync with the actual code base. Otherwise you need to support multiple versions in your stubs and this also requires a lot of effort to keep them in sync.

Comment (zero323) [reply python/pyspark/sql/functions.py:2396]: Pre 3.6 wasn't an issue anyway.. In practice, given the size of the project and overall complexity of annotations, stubs seem to be a somewhat better choice anyway (avoid cyclic imports or certain `Generic` pitfalls for starters, not being stuck in legacy Python support). This means that potential migration might be relatively easy.

My biggest concern is that over four years I haven't seen any active Spark contributors willing to help with maintenance (and that's despite gentle  pings). Additionally annotation ecosystem is still evolving ‒ so having a project that can iterate faster than annotated codebase is not a bad thing. 

The biggest advantage of maintaining annotations alongside actual codebase is that it requires conscious choices of signatures ‒ not every type of signature can be annotated in a meaningful way and some can result in exploding overloads. But it doesn't really remove the maintenance overhead ‒ for example changes in the ML API, mostly invisible for the final user, wreaked havoc, as it is the most vulnerable part of code (a lot of generics there). 

Anyway... That's just my 2 cents. If you reanimate the discussion. please ping me ‒ I'd like to see how it goes. TIA

File: python/pyspark/ml/tuning.py
Comment (Fokko) [reply python/pyspark/ml/tuning.py]: At the same time, this would change the public API

File: python/pyspark/sql/functions.py
Comment (Fokko) [reply python/pyspark/sql/functions.py:2396]: I'm happy to work on the annotations, but would be great to have some support from the committers as well. I'm planning to integrate mypy in the CI as well, so we can enforce the type checking in the main repository as well. Personally, I don't see any harm in adding the types over time, and work towards coverage, starting with the public functions because they add the most value.

Comment (zero323) [reply python/pyspark/sql/functions.py:2396]: > Personally, I don't see any harm in adding the types over time, and work towards coverage,

The biggest problem is the resolution (https://bugs.python.org/issue39168 seems to be fixed now). As far as I am aware it is all-or-nothing at the module level (assuming `partial` mode), so working over time will likely degrade experience of the users that already use annotations.

And it seems somewhat unnecessary, considering that you could migrate stubs into main repo with a single copy, if there was a will to maintain the result. Inline variant would be trickier (and there is still concern of https://bugs.python.org/issue39168).

Comment (zero323) [reply python/pyspark/sql/functions.py:2396]: How about continue the discussion [here](Given a discussion related to <a href="https://github.com/apache/spark/pull/29122">SPARK-32320 PR</a> I'd like to resurrect this thread. Is there any interest in migrating annotations to the main repository?) for a better visibility?

Comment (Fokko) [reply python/pyspark/sql/functions.py:2396]: The bug is fixed in 3.6, so that shouldn't be a problem. Better to move this to the maillist indeed, thanks!

Comment (Fokko) [reply python/pyspark/sql/functions.py]: I've removed all the inline type annotations :)

**Comment by Fokko** [comment]
Fixed the voilations:
```
./python/pyspark/ml/regression.py:1743:40: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
                 quantileProbabilities=list([0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]),
                                       ^
./python/pyspark/ml/regression.py:1761:41: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
                  quantileProbabilities=list([0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]),
                                        ^
./python/pyspark/ml/tuning.py:511:46: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, bestModel, avgMetrics=[], subModels=None):
                                             ^
./python/pyspark/ml/tuning.py:871:53: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, bestModel, validationMetrics=[], subModels=None):
                                                    ^
./python/pyspark/resource/profile.py:35:63: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, _java_resource_profile=None, _exec_req={}, _task_req={}):
                                                              ^
./python/pyspark/resource/profile.py:35:77: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, _java_resource_profile=None, _exec_req={}, _task_req={}):
                                                                            ^
./python/pyspark/sql/functions.py:2479:36: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def from_json(col, schema, options={}):
                                   ^
./python/pyspark/sql/functions.py:2527:26: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def to_json(col, options={}):
                         ^
./python/pyspark/sql/functions.py:2567:34: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def schema_of_json(json, options={}):
                                 ^
./python/pyspark/sql/functions.py:2597:32: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def schema_of_csv(csv, options={}):
                               ^
./python/pyspark/sql/functions.py:2623:25: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def to_csv(col, options={}):
                        ^
./python/pyspark/sql/functions.py:2934:35: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def from_csv(col, schema, options={}):
                                  ^
./python/pyspark/sql/avro/functions.py:29:47: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
def from_avro(data, jsonFormatSchema, options={}):
                                              ^
./dev/sparktestsupport/modules.py:34:97: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
    def __init__(self, name, dependencies, source_file_regexes, build_profile_flags=(), environ={},
                                                                                                ^
14    B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
14
1
```

Furthermore:
- Added `flake8-bugbear` to the CI. This is a well known additional set of rules next to flake8: https://github.com/pycqa/flake8-bugbear
- Added the `B006` rule, which is the mutable arguments rule :)

File: python/pyspark/sql/avro/functions.py
Comment (zero323) [inline python/pyspark/sql/avro/functions.py:29]: We should change annotations from

```python
def from_avro(
    data: ColumnOrName, jsonFormatSchema: str, options: Dict[str, str] = ...
) -> Column: ..
```

to

```python
def from_avro(
    data: ColumnOrName, jsonFormatSchema: str, options: Optional[Dict[str, str]] = ...
) -> Column: ..
```

File: python/pyspark/sql/functions.py
Comment (zero323) [inline python/pyspark/sql/functions.py:2481]: Same as for `sql.avro.functions`, we'll need here `Optional[Dict[str, str]]` (applies also to the remaining functions).

File: python/pyspark/ml/tuning.py
Comment (zero323) [inline python/pyspark/ml/tuning.py:511]: Annotation

https://github.com/apache/spark/blob/46ad325e56abd95c0ffdbe64aad78582da8c725d/python/pyspark/ml/tuning.pyi#L107

should change to  `avgMetrics: Optional[List[float]] = ...,`

File: python/pyspark/ml/tuning.py
Comment (zero323) [inline python/pyspark/ml/tuning.py:871]: Annotation

https://github.com/apache/spark/blob/46ad325e56abd95c0ffdbe64aad78582da8c725d/python/pyspark/ml/tuning.pyi#L174

should change to `validationMetrics: Optional[List[float]] = ...,`

File: python/pyspark/ml/regression.py
Comment (zero323) [inline python/pyspark/ml/regression.py:1745]: This is still a mutable default, isn't it? Shouldn't `DEFAULT_QUANTILE_PROBABILITIES` by a `tuple`, if we want to be consistent?

File: python/pyspark/ml/regression.py
Comment (zero323) [inline python/pyspark/ml/regression.py:1659]: Side note ‒ I remember this was included for Scala parity (SPARK-32310), but isn't that a dead code, give we use the same default in estimator and model constructors?

File: python/pyspark/sql/avro/functions.py
Comment (Fokko) [reply python/pyspark/sql/avro/functions.py:29]: Thanks. Just curious, how can we check this? Mypy has been added, and I would expect that this would be caught.

File: python/pyspark/ml/regression.py
Comment (Fokko) [reply python/pyspark/ml/regression.py:1745]: Bugbear doesn't complain, but you're right:
```
MacBook-Pro-van-Fokko:spark fokkodriesprong$ python3
Python 3.6.6 (v3.6.6:4cf1f54eb7, Jun 26 2018, 19:50:54) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import List
>>> 
>>> DEFAULT_QUANTILE_PROBABILITIES = [0.01, 0.1, 0.99]
>>> 
>>> def mutate_arguments(mutate: List = DEFAULT_QUANTILE_PROBABILITIES):
...   mutate.append(22)
...   print(mutate)
... 
>>> mutate_arguments()
[0.01, 0.1, 0.99, 22]
>>> mutate_arguments()
[0.01, 0.1, 0.99, 22, 22]
>>> mutate_arguments()
[0.01, 0.1, 0.99, 22, 22, 22]
>>> mutate_arguments()
[0.01, 0.1, 0.99, 22, 22, 22, 22]
```

File: python/pyspark/ml/regression.py
Comment (Fokko) [reply python/pyspark/ml/regression.py:1659]: Yes, you are correct. The code isn't being used. I've removed the default argument.

Comment (zero323) [reply python/pyspark/ml/regression.py:1659]: Sorry for the confusion ‒ I just wanted to point it out to see if I missed something, but I don't think we should change it here, as this will be a breaking change.

File: python/pyspark/sql/avro/functions.py
Comment (zero323) [reply python/pyspark/sql/avro/functions.py:29]: >  Mypy has been added, and I would expect that this would be caught.

As far as I recall this behavior depends on the MyPy flags. In particular see [Disabling strict optional checking](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#no-strict-optional).

So adding `no_implicit_optional` to `mypy.ini` should catch this.

Comment (zero323) [reply python/pyspark/sql/avro/functions.py:29]: FYI I am trying to establish optimal set of flags for mypy config, but it is a bit slowish work ‒ unlike standalone stubs, we have to correct for internal modules, tests and such, and I'd prefer to avoid atomic options (ignoring all errors there).

File: python/pyspark/ml/regression.py
Comment (Fokko) [reply python/pyspark/ml/regression.py:1659]: I've removed the default argument in the dead code.

File: python/pyspark/sql/avro/functions.py
Comment (Fokko) [reply python/pyspark/sql/avro/functions.py:29]: Thanks! 👍  Yes, definitely a process that we have to go through. Let's break this up into smaller pieces.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @huaxingao and @zhengruifeng FYI

File: python/pyspark/ml/regression.py
Comment (Fokko) [reply python/pyspark/ml/regression.py:1659]: Yes, since there is only one place where this is used, let's revert this.

File: python/pyspark/sql/functions.py
Comment (zero323) [inline python/pyspark/sql/functions.py:3476]: Seems like we still have to modify a few annotations, right?

Probably something like [functions.pyi.patch.txt](https://github.com/apache/spark/files/5583207/functions.pyi.patch.txt)

Comment (Fokko) [reply python/pyspark/sql/functions.py:3476]: Good catch, I've added them

Comment (Fokko) [reply python/pyspark/sql/functions.py:3476]: Not sure why implicit optional is still allowed 🤔

**Comment by shaneknapp** [comment]
test this please

**Review by zero323 — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]
@Fokko please update the description of this PR.

**Comment by Fokko** [comment]
@zhengruifeng done! :)

**Comment by HyukjinKwon** [comment]
Just for the record and FWIW I don't mind merging this PR. I will leave it to other committers here.

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@zhengruifeng the tests are broken globally for an unknown reason. It's being investigated at https://github.com/apache/spark/pull/30645 and it's hardly related to the current change. I think it's fine to merge if the test results bother you.

**Comment by zhengruifeng** [comment]
@HyukjinKwon Thanks for your explanation!

**Comment by zhengruifeng** [comment]
Merged to master, thanks all!

**Pull request closed without merging**
